File: lib/monoboly_deal_web/channels/player_channel.ex
defmodule MonobolyDealWeb.PlayerChannel do
  use MonobolyDealWeb, :channel

  alias MonobolyDeal.Game.Server

  def join("players:" <> _player_name, _params, socket) do
    {:ok, socket}
  end

  def handle_in("hand_dealt", %{"game_name" => game_name}, socket) do
    case Server.game_pid(game_name) do
      pid when is_pid(pid) ->
        players_hand = Server.get_hand(game_name, socket.assigns.current_player)

        push(socket, "player_hand", %{hand: players_hand})
        {:noreply, socket}

      nil ->
        {:reply, {:error, %{reason: "Player does not exist"}}, socket}
    end
  end
end

File: assets/js/communication/PlayerChannel.ts
import {Channel, Socket} from "phoenix"
import {Card} from "../models/Card"
import {CardSerializer} from "../serializers/CardSerializer"

export class PlayerChannel {
  private channel: Channel
  private cardSerializer: CardSerializer

  constructor(playerName: string, socket: Socket) {
    this.channel = socket.channel("players:" + playerName)
    this.cardSerializer = new CardSerializer()

    this.channel
      .join()
      .receive("ok", (response: any) => {
        console.log("player channel joined", {response})
      })
      .receive("error", (reason: any) => {
        console.log("player channel join failed", {reason})
      })
  }

  public onHandUpdated(callBack: (hand: Card[]) => void) {
    this.channel.on("player_hand", (response: { hand: object[] }) => {
      if (response && response.hand) {
        const cards = response.hand.map((cardJson: object) => {
          return this.cardSerializer.deserialize(cardJson)
        })

        callBack(cards)
      }
    })
  }
}
